[nmstate-0.3] bridge: ignore unmanaged ports always

ffmancera opened the pull request:
Nmstate will ignore unmanaged ports. If an existing unmanaged port of a
bridge is not being included in the desired state bridge port list, it
will not be removed. Basically, Nmstate will not be able to remove
unmanaged ports from a bridge.

Ref: https://bugzilla.redhat.com/1932247

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

cathay4t commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -351,7 +351,7 @@ def state_to_edit(self):
         return self._cur_ifaces
 
-    def _remove_unknown_interface_type_slaves(self):
+    def _remove_unknown_unmanaged_interface_type_slaves(self):
```
unmanged is nm specific.

ffmancera replied:
Removed :-)

cathay4t commented on libnmstate/nm/applier.py:
```diff
@@ -90,7 +92,10 @@ def apply_changes(context, net_state, save_to_disk):
-        lambda s: s.get(Interface.STATE) != InterfaceState.ABSENT,
+        lambda s: s.get(Interface.STATE) == InterfaceState.UP
+        or (
+            s.get(Interface.STATE) == InterfaceState.DOWN and s.get(IS_DESIRED)
```
~~This clearly will break a lot in CI.~~

Why not use `not nm_dev.get_managed() and is_desired`

ffmancera replied:
Not needed anymore. I've used your suggestions

cathay4t commented:
@ffmancera Can you also try:

```patch
diff --git a/libnmstate/ifaces/ifaces.py b/libnmstate/ifaces/ifaces.py
index 5f1f548..cddeace 100644
--- a/libnmstate/ifaces/ifaces.py
+++ b/libnmstate/ifaces/ifaces.py
@@ -193,7 +193,7 @@ class Ifaces:
         if not defiend in desire state
         """
         for iface in self._ifaces.values():
-            if iface.is_up and iface.is_master:
+            if iface.is_desired and iface.is_up and iface.is_master:
                 cur_iface = self.current_ifaces.get(iface.name)
                 for slave_name in iface.slaves:
                     if cur_iface and slave_name in cur_iface.slaves:
diff --git a/libnmstate/nm/applier.py b/libnmstate/nm/applier.py
index cd319ff..03c282f 100644
--- a/libnmstate/nm/applier.py
+++ b/libnmstate/nm/applier.py
@@ -80,7 +80,16 @@ def apply_changes(context, net_state, save_to_disk):
     _preapply_dns_fix(context, net_state)
     _mark_nm_external_subordinate_changed(context, net_state)

-    ifaces_desired_state = net_state.ifaces.state_to_edit
+    ifaces_desired_state = []
+    for iface in net_state.ifaces.values():
+        nm_dev = context.get_nm_dev(iface.name)
+        if nm_dev and not nm_dev.get_managed() and not iface.is_desired:
+            # We don't change NM.Device from unmanaged to managed unless
+            # been asked explicitly in desire state
+            continue
+        if iface.is_changed or iface.is_desired:
+            ifaces_desired_state.append(iface.to_dict())
+
     ifaces_desired_state.extend(
         _create_proxy_ifaces_desired_state(ifaces_desired_state)
     )
```

ffmancera committed:
nm.applier: ignore not desired unmanaged ifaces

Nmstate will ignore not desired unmanaged interfaces. If an existing
unmanaged port of a bridge is not being included in the desired state
bridge port list, it will not be removed. Basically, Nmstate will not be
able to remove unmanaged ports from a bridge.

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

ffmancera commented:
> @ffmancera Can you also try:
> 
> ```diff
> diff --git a/libnmstate/ifaces/ifaces.py b/libnmstate/ifaces/ifaces.py
> index 5f1f548..cddeace 100644
> --- a/libnmstate/ifaces/ifaces.py
> +++ b/libnmstate/ifaces/ifaces.py
> @@ -193,7 +193,7 @@ class Ifaces:
>          if not defiend in desire state
>          """
>          for iface in self._ifaces.values():
> -            if iface.is_up and iface.is_master:
> +            if iface.is_desired and iface.is_up and iface.is_master:
>                  cur_iface = self.current_ifaces.get(iface.name)
>                  for slave_name in iface.slaves:
>                      if cur_iface and slave_name in cur_iface.slaves:
> diff --git a/libnmstate/nm/applier.py b/libnmstate/nm/applier.py
> index cd319ff..03c282f 100644
> --- a/libnmstate/nm/applier.py
> +++ b/libnmstate/nm/applier.py
> @@ -80,7 +80,16 @@ def apply_changes(context, net_state, save_to_disk):
>      _preapply_dns_fix(context, net_state)
>      _mark_nm_external_subordinate_changed(context, net_state)
> 
> -    ifaces_desired_state = net_state.ifaces.state_to_edit
> +    ifaces_desired_state = []
> +    for iface in net_state.ifaces.values():
> +        nm_dev = context.get_nm_dev(iface.name)
> +        if nm_dev and not nm_dev.get_managed() and not iface.is_desired:
> +            # We don't change NM.Device from unmanaged to managed unless
> +            # been asked explicitly in desire state
> +            continue
> +        if iface.is_changed or iface.is_desired:
> +            ifaces_desired_state.append(iface.to_dict())
> +
>      ifaces_desired_state.extend(
>          _create_proxy_ifaces_desired_state(ifaces_desired_state)
>      )
> ```

I've used this suggestions. They work fine, I do not like it with the original 0.3 applier design but well, I do not think there is a better solution.

Please, check out the new patch.
